keps/sig-auth/20190607-certificates-api.md

```diff
@@ -0,0 +1,285 @@
+   a third party masquerading as an authorized subject.
+1. The subject of the CSR is authorized to act in the requested context. This
+   addresses the threat of an undesired subject joining the cluster.
+
+Iff these two requirements are met, the approver should approve the CSR and
+otherwise should deny the CSR. Once the CSR is approved a Certificate Authority
+(referred to in this design as the signer) should construct a certificate from
+the CSR and return the certificate to the requester.
+
+The Certificates API provides a generic means of communication exposed via the
+Kubernetes resource model over which a certificate requestor, approver and
+signer can interact.
+
+## Design Details
+
+A client requesting a certificate post a CertificateSigningRequest to the
+Certificates API. The client may only provide the encoded [Certificate
+Request](https://tools.ietf.org/html/rfc2986) and usages of the certificate in
+the spec (and the standard object metadata) on the initial creation of the
+CertificateSigningRequest. The kube-apiserver also asserts authentication
+attributes of the requestor in the CertificateSigningRequest spec before
+committing it to storage so that they can be used later during CSR approval. The
+information contained in the spec is immutable after the request is created.
+
+The approver updates approval status of the CertificateSigningRequest via the
+CertificateSigningRequestStatus. The approval condition can only be updated via
+the `/approve` subresource allowing approval permission to be authorized
+independently of other operations on the CertificateSigningRequest.
+
+Contingent on approval, the signer posts a signed certificate to the status. The
+certificate field of the status can only be updated via the `/status`
+subresource allowing signing permission to be authorized independently of other
+operations on the CertificateSigningRequest.
+
+The API is designed to support the standard asynchronous controller model of
+Kubernetes where the approver and signer act as independent controllers of the
+Certificates API. Since issuance is asynchronous, the approver can perform
+out-of-band verification of the CSR before making an authorization decision.
+
+The approver is designed to be explicitly independent of the signer. This
+separates concerns of authorization and certificate minting and also signer to
+operate in a separate, more restrictive environment than the approver. This is
+typical of many PKI architectures.
+
+### Sequence of an Issuance
+
+A typical successful issuance proceeds as follows.
+
+![CSR](/keps/sig-auth/csr.png)
+
+1. The requestor generates a private key, builds a certificate signing request,
+   and Submits the certificate request to the Kubernetes certificates API.
+1. The approver controller Observes the newly submitted request, Validates and
+   authorizes the request And if all goes well, approves the request
+1. The signer observes the approval and mints a new certificate.
```

Comment:
"The signer observes the approval, mints a new certificate and stores it in the `CertificateSigningRequest.Status.Certificate` field"